### PR TITLE
Clean-up doc

### LINKED ISSUE #9: clean up task definitions and results
delete unnecessary / non official plots, and clarify metrics and definitions. make results page match the tasks precisely and vice versa. we can later iterate on details of stepsize-schedule etc, once we have one fixed setup of task and results

follow-up to https://github.com/mlbench/mlbench-docs/pull/8

### PR DESCRIPTION
**Pull request by ehoelzl**
Update to the project documentation:
- Remove authors as instructed
- Rename tasks and group them
- Verify and change the hyper-parameters

Closes #9

### CODE REVIEW

File: benchmark-tasks.rst [hunk 1]
```diff
@@ -33,11 +33,25 @@ precise comparison of most state-of-the-art algorithms, frameworks, and hardware
 
 For each task, we provide a reference implementation, as well as results for different systems.
 
-1 Image Classification Tasks
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Task 0: Communication Backend
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: Task 0: Communication Backend Raw Performance

File: benchmark-tasks.rst [hunk 2]
```diff
@@ -33,11 +33,25 @@ precise comparison of most state-of-the-art algorithms, frameworks, and hardware
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-1a. Image Classification (ResNet, CIFAR-10)
-###########################################
+This task consists of benchmarking the communication backends for different frameworks and operations.
+
+0.a PyTorch All-reduce
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: just all-reduce, doesn't need to be pytorch in the task (but the ref impl is pytorch)

File: benchmark-tasks.rst [hunk 3]
```diff
@@ -48,8 +62,7 @@ Image classification is one of the most important problems in computer vision an
     `this paper <https://arxiv.org/abs/1512.03385>`_.
     The second version (m2) is based on the ResNets defined `here
     <https://arxiv.org/abs/1603.05027>`_.
-    For each version we have the network implementations
-    with 20, 32, 44, and 56 layers.
+    For this benchmark implementation, we use 20 layer ResNet called ResNet20.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: typo 'a 20 lay...' (and maybe add a reference if there is an obvious one)

File: benchmark-tasks.rst [hunk 4]
```diff
@@ -71,12 +84,12 @@ Image classification is one of the most important problems in computer vision an
     - maximum epochs: 164
     - learning rate
 
-      + learning rate per sample :math:`\eta` : 0.1 / 256
-      + decay: similar to `Deep Residual Learning for Image Recognition`_, we reduce learning rate by 1/10 at the 82-th and 109-th epoch.
+      + learning rate :math:`\eta` : 0.02
+      + decay: We reduce the learning rate when a plateau in the validation loss is reached for 2 consecutive epochs
       + scaling and warmup: apply ``linear scaling rule`` mentioned in :cite:`goyal2017accurate`. The learning rate per worker is scaled from
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: learning rate per worker, should be same as global learning rate, no? try to avoid confusion. if per worker can be skipped let's remove it

File: benchmark-tasks.rst [hunk 5]
```diff
@@ -91,24 +104,26 @@ Implementation details:
 
 #. **Environments for Scaling Task**
     For the scaling task, we use `n1-standard-4 <https://cloud.google.com/compute/pricing>`_ type instances with 50GB disk size.
-    There is only one worker per node; each worker uses 2.5 cpus. The bandwidth between two nodes is around 7.5Gbit/s.
+    There is only one worker per node; each worker uses 3 cpus. The bandwidth between two nodes is around 7.5Gbit/s.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: upper case cpus. and scaling result should first be described for GPUs, no? CPUs can be mentioned after it

File: benchmark-tasks.rst [hunk 6]
```diff
@@ -118,32 +133,17 @@ TODO
...
-        We train the model with different batch sizes ([1,..,1000]) and in the end we select the batch size
-        that enables the trained model to reach to 89% accuracy on the validation set in less time. we use
-        80% of the dataset to train the model, and the remaining 20% is used as the validation set.
-.. [2] :math:`\alpha` is tuned for each cluster size separately. To do so, we use 80% of the dataset to train
-        the model, and the remaining 20% is used as the validation set. We do a grid search to find the best
-        value for alpha: for each value in the grid ([0.001,..,1000]), the model is trained until it reaches
-        to 89% accuracy on the validation set. Finally, we select the value that enables the model to reach
-        the target accuracy value faster.
+    - optimizer: ``CentralizedSGD(momentum=0, nesterov=False, weight_decay=0, dampening=0, by_layer=False)``
+    - regularization rate: :math:`L1=0, L2 = 0.0000025`
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: looks good.  regularization rate -> regularization parameters

File: benchmark-tasks.rst [hunk 7]
```diff
@@ -152,23 +152,106 @@ TODO
 
 #. **Environments for Scaling Task**
     For the scaling task, we use `n1-standard-4 <https://cloud.google.com/compute/pricing>`_ type instances with 50GB disk size.
-    There is only one worker per node; each worker uses 2.5 cpus. The bandwidth between two nodes is around 7.5Gbit/s.
-    Openmpi is used for communication. No accelerators are used for this task.
+    There is only one worker per node; each worker uses 3 cpus. The bandwidth between two nodes is around 7.5Gbit/s.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: CPUs (do a global search and replace)

File: benchmark-tasks.rst [hunk 8]
```diff
@@ -152,23 +152,106 @@ TODO
...
+Task 4: Machine Translation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _benchmark-task-4a:
+4.a LSTM, WMT16 EN-DE
+"""""""""""""""""""""
+
+
+#. **Model**
+    We benchmark the `GNMT <https://arxiv.org/abs/1609.08144>`_, which follows the sequence-to-sequence learning framework,
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: the GNMT machine translation model

File: benchmark-tasks.rst [hunk 9]
```diff
@@ -152,23 +152,106 @@ TODO
...
+    and uses stacked residual LSTM connections in the encoder and decoder modules. The residual connections allow
+    for deeper stacked LSTM layers, as without residuals, the stack typically suffer from
+    vanishing/exploding gradients when too many layers are used.
```
Comment (martinjaggi) [inline benchmark-tasks.rst:195]: potentially add a reference (paper or MLperf), to be as reproducible as possible?

File: benchmark-tasks.rst [hunk 10]
```diff
@@ -152,23 +156,109 @@ TODO
...
+
+#. **Dataset**
+    The `WMT-16 <http://www.statmt.org/wmt16/metrics-task/>`_
+    dataset containing a set of translated sentences from multiple languages.
+    We exclusively use English-German translation from this dataset.
+
+
+#. **Training Algorithm**
+    We use Distributed Adam as the optimizer. Before each weight update, gradients on all workers are average using an `all_reduce` operation.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: add reference if possible

File: benchmark-tasks.rst [hunk 11]
```diff
@@ -152,23 +156,109 @@ TODO
...
+        - Before each backward pass, inflate the loss by `loss_scaling` (in `float16`) to avoid underflows
+        - Before weight update, deflate gradients by `loss_scaling` (in `float32`) to keep precision
+        - Clip gradient norm to be `grad_clip`
+        - Check if gradient norm is `nan` or `inf` (in `float16`). If True, `loss_scale = loss_scale / scale_factor`.
+          If False, update weights.
+        - If after `scale_window` updates, no overflow/underflow detected, `loss_scale = loss_scale * scale_factor`
+
+#. **Selection of Framework & Systems**
+    We currently only have this reference implementation in PyTorch. For the systems, kubernetes allows easy transferability of our code.
+    While initial results reported are from google kubernetes engine, AWS will be supported very soon.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: G

**Review by Panaetius — APPROVED** [review]

**Review by martinjaggi — CHANGES_REQUESTED** [review]

File: benchmark-tasks.rst (hunk 5 above)
Comment (ehoelzl) [reply benchmark-tasks.rst]: Maybe remove completely this part and keep the hardware mentioned in the Results ?

**Comment by martinjaggi** [comment]
sounds good

anyway it might make sense for each task to have the main things first,
followed by a small 'fine print'/technical details section?

File: benchmark-tasks.rst (hunk 9 above)
Comment (ehoelzl) [reply benchmark-tasks.rst:195]: Paper is referenced right above. I can add MLPerf's implementation as reference too if needed

**Comment by martinjaggi** [comment]
ok. ideally we'd use the same referencing format for all papers (also the
goyal one for LR scaling which is currently at the end or in a footnote.)
whatever is easiest.

**Comment by martinjaggi** [comment]
in the results stucture, can you fuse Machine Type + GPU Type into one item, maybe called `System Hardware` or similar?

**Review by martinjaggi — APPROVED** [review]

**Comment by martinjaggi** [comment]
can we merge this? maybe has to be rebased to latest version?
